internal/daemon/api_files.go

```diff
@@ -0,0 +1,627 @@
+	UserID      int    `json:"user-id"`
+	User        string `json:"user"`
+	GroupID     int    `json:"group-id"`
+	Group       string `json:"group"`
+}
+
+func writeFiles(body io.Reader, boundary string) Response {
+	// Read metadata part (field name "request").
+	mr := multipart.NewReader(body, boundary)
+	part, err := mr.NextPart()
+	if err != nil {
+		return statusBadRequest("cannot read request metadata: %v", err)
+	}
+	if part.FormName() != "request" {
+		return statusBadRequest(`metadata field name must be "request", not %q`, part.FormName())
+	}
+
+	// Decode metadata about files to write.
+	var payload struct {
+		Action string           `json:"action"`
+		Files  []writeFilesItem `json:"files"`
+	}
+	decoder := json.NewDecoder(part)
+	if err := decoder.Decode(&payload); err != nil {
+		return statusBadRequest("cannot decode request metadata: %v", err)
+	}
+	if payload.Action != "write" {
+		return statusBadRequest(`multipart action must be "write", not %q`, payload.Action)
+	}
+	if len(payload.Files) == 0 {
+		return statusBadRequest("must specify one or more files")
+	}
+	infos := make(map[string]writeFilesItem)
+	for _, file := range payload.Files {
+		infos[file.Path] = file
+	}
+
+	errors := make(map[string]error)
+	for i := 0; ; i++ {
+		part, err = mr.NextPart()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return statusBadRequest("cannot read file part %d: %v", i, err)
+		}
+		if !strings.HasPrefix(part.FormName(), "file:") {
+			return statusBadRequest(`field name must be in format "file:/path", not %q`, part.FormName())
+		}
+		p := part.FormName()[len("file:"):]
+		info, ok := infos[p]
+		if !ok {
+			return statusBadRequest("no metadata for path %q", p)
+		}
+		errors[p] = writeFile(info, part)
+		part.Close()
+	}
+
+	// Build list of results with any errors.
+	result := make([]fileResult, len(payload.Files))
+	var firstErr error
+	for i, file := range payload.Files {
+		err, ok := errors[file.Path]
+		if !ok {
+			// Ensure we wrote all the files in the metadata.
+			err = fmt.Errorf("no file content for path %q", file.Path)
+		}
+		if err != nil && firstErr == nil {
+			firstErr = err
+		}
+		result[i] = fileResult{
+			Path:  file.Path,
+			Error: fileErrorToResult(err),
+		}
+	}
+	return syncResponseWithError(result, firstErr)
+}
+
+func writeFile(item writeFilesItem, source io.Reader) error {
+	if !path.IsAbs(item.Path) {
+		return nonAbsolutePathError(item.Path)
+	}
+
+	// Create parent directory if needed
+	if item.MakeDirs {
+		err := os.MkdirAll(path.Dir(item.Path), 0o755)
+		if err != nil {
+			return fmt.Errorf("creating directory: %w", err)
+		}
+	}
+
+	// Create file and write contents to temporary file.
+	perm, err := parsePermissions(item.Permissions, 0o644)
+	if err != nil {
+		return err
+	}
+	tempPath := item.Path + ".pebble-temp"
+	f, err := os.OpenFile(tempPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, perm)
+	if err != nil {
+		return fmt.Errorf("creating file: %w", err)
```

Comment:
Depending on where this is presented and what happens in "err", that message might be unclear as it sounds like an action is taking place right now.  The traditional `"cannot create file: %w"`, would be safer. Same for the cases below.

Reply:
Good call, updated.